sys/dev_vga.c

```diff
@@ -102,10 +103,43 @@ static int dev_vga_lookup(vnode_t *v, const char *name, vnode_t **res) {
   return ENOENT;
 }
 
+static void uiomove_vga_dirent(const char *name, uio_t *uio) {
+  dirent_t *dir = NULL;
+  dir = (dirent_t *)kmalloc(M_VFS, _DIRENT_RECLEN(dir, strlen(name)), M_ZERO);
+  dir->d_namlen = strlen(name);
+  dir->d_reclen = _DIRENT_RECLEN(dir, dir->d_namlen);
+  memcpy(dir->d_name, name, dir->d_namlen + 1);
+  uiomove(dir, dir->d_reclen, uio);
+  kfree(M_VFS, dir);
+}
+
+static int dev_vga_readdir(vnode_t *v, uio_t *uio) {
+  int start_offset = uio->uio_offset;
+  dirent_t *dir;
+  int fb_offset = 0;
+  int palette_offset = fb_offset + _DIRENT_RECLEN(dir, strlen("fb"));
+  int videomode_offset =
+    palette_offset + _DIRENT_RECLEN(dir, strlen("palette"));
+
+  if (uio->uio_offset == fb_offset &&
+      uio->uio_resid >= _DIRENT_RECLEN(dir, strlen("fb")))
+    uiomove_vga_dirent("fb", uio);
+
+  if (uio->uio_offset == palette_offset &&
+      uio->uio_resid >= _DIRENT_RECLEN(dir, strlen("palette")))
+    uiomove_vga_dirent("palette", uio);
+
+  if (uio->uio_offset == videomode_offset &&
+      uio->uio_resid >= _DIRENT_RECLEN(dir, strlen("videomode")))
+    uiomove_vga_dirent("palette", uio);
```

Comment:
Please don't mix code and data. You've made a classic mistake here (copy & paste). Put data into separate array and write a procedure that will transform `vga_dir` to `uio`-able buffer.